src/wccp2.cc

```diff
@@ -1476,6 +1486,9 @@ wccp2HandleUdp(int sock, void *)
 
                 SetField(cache_identity, ptr, router_view_header, router_view_size,
                          "malformed packet (truncated router view info cache w/o assignment hash)");
+                if (cache_identity->bits & 6 >> 1 != 0) {
+                    fatalf("Assignment type mismatch in router view");
```

Comment:
This needs to throw instead. The core problem behind CVE 2021-28116 was letting values supplied by routers abort Squid.

Reply:
My understanding of the problem was that a broken/fake packet could access memory beyond its size.
Initially I used Must3 like at other places, but the outcome was not good (throw after some struct updated caused list of caches corrupted leading to segfault during processing of next packet) or infinite loop of negotiation attempt and ignored message, like the current situation.

I preferred to let the cache stop entirely over continuing with wccp disabled. It should happen during initial setup only. If that is not a good approach, I would also reinitialize wccp completely and start negotiation from scratch.

Reply:
The code parsing received traffic should not reach fatal(). In the vast majority of cases, when throwing breaks things, it is not throwing that should be blamed -- the bug is elsewhere (and should be fixed elsewhere). Moreover, IIRC, this code already throws, even before this PR changes. Thus, even if this PR avoids throwing, the post-throwing problems you have observed will probably continue to exist; you are not really eliminating them.

> It should happen during initial setup only

If not today, then tomorrow that "initial setup" event might happen ten days after the Squid instance has started (e.g., as a result of a reconfiguration).

> I would also reinitialize wccp prior exception and force negotiation from scratch.

Please handle exceptions when/after catching them, not when/prior to throwing them. Adding a `try/catch` may be necessary -- I have not checked.

Reply:
Ah. The design of the logic is that a thrown exception when handling a packet aborts handling only that packet. Any information taken out of it should only be saved after validation of that piece of information.

Since WCCP sends a bunch of settings Squid is able to accept one piece (eg return method) while rejecting others (eg router list) provided the one used is valid regardless of the rejected piece. When there is any dependency situation between packet values they should be kept in local variables until it is all validated. Not added to the operational config as separate updates.